hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java

```diff
@@ -0,0 +1,144 @@
+import com.hazelcast.core.ExecutionCallback;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.cluster.Versions;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+import com.hazelcast.spi.merge.SplitBrainMergePolicyProvider;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+
+/**
+ * Merges data structures which have been collected via an {@link AbstractContainerCollector}.
+ *
+ * @param <C> container of the data structure
+ */
+public abstract class AbstractContainerMerger<C> implements Runnable {
+
+    private static final long TIMEOUT_FACTOR = 500;
+    private static final long MINIMAL_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(5);
+
+    protected final AbstractContainerCollector<C> collector;
+
+    private final Semaphore semaphore = new Semaphore(0);
+    private final ExecutionCallback<Object> mergeCallback = new ExecutionCallback<Object>() {
+        @Override
+        public void onResponse(Object response) {
+            semaphore.release(1);
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            logger.warning("Error while running " + getLabel() + " merge operation: " + t.getMessage());
+            semaphore.release(1);
+        }
+    };
+
+    private final ILogger logger;
+    private final ClusterService clusterService;
+    private final OperationService operationService;
+    private final SplitBrainMergePolicyProvider splitBrainMergePolicyProvider;
+
+    private int operationCount;
+
+    protected AbstractContainerMerger(AbstractContainerCollector<C> collector, NodeEngine nodeEngine) {
+        this.collector = collector;
+        this.logger = nodeEngine.getLogger(AbstractContainerMerger.class);
+        this.clusterService = nodeEngine.getClusterService();
+        this.operationService = nodeEngine.getOperationService();
+        this.splitBrainMergePolicyProvider = nodeEngine.getSplitBrainMergePolicyProvider();
+    }
+
+    @Override
+    public final void run() {
+        // we cannot merge into a 3.9 cluster, since not all members may understand the new merge operation
+        // RU_COMPAT_3_9
+        if (clusterService.getClusterVersion().isLessThan(Versions.V3_10)) {
+            logger.info("Cluster needs to run version " + Versions.V3_10 + " to merge " + getLabel() + " instances");
+            return;
+        }
+        int valueCount = collector.getMergingValueCount();
+        if (valueCount == 0) {
+            return;
+        }
+
+        runInternal();
+
+        assert operationCount > 0 : "No merge operations have been invoked in AbstractContainerMerger";
+
+        try {
+            long timeoutMillis = Math.max(valueCount * TIMEOUT_FACTOR, MINIMAL_TIMEOUT_MILLIS);
```

Comment:
I've added this, so we wait at least 5 seconds for all merge operations to finish (even for a single one). If the operations finish faster, we won't wait any longer. I just wanted to prevent an eagerly sent MERGED lifecycle event. Maybe we should even define a longer minimum timeout value?